src/main/java/net/imglib2/parallel/DefaultTaskExecutor.java

```diff
@@ -0,0 +1,183 @@
+ * Jean-Yves Tinevez and Michael Zinsmaier.
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+package net.imglib2.parallel;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * A {@link TaskExecutor} that wraps around a given {@link ExecutorService}.
+ */
+public class DefaultTaskExecutor implements TaskExecutor
+{
+
+	private final ExecutorService executorService;
+
+	public DefaultTaskExecutor( ExecutorService executorService )
+	{
+		this.executorService = executorService;
+	}
+
+	@Override
+	public ExecutorService getExecutorService()
+	{
+		return executorService;
+	}
+
+	@Override
+	public int getParallelism()
+	{
+		if ( executorService instanceof ForkJoinPool )
+			return ( ( ForkJoinPool ) executorService ).getParallelism();
+		else if ( executorService instanceof ThreadPoolExecutor )
+			return Math.max( 1, ( ( ThreadPoolExecutor ) executorService ).getCorePoolSize() );
+		else if ( executorService instanceof ForkJoinExecutorService )
+			return ( ( ForkJoinExecutorService ) executorService ).getParallelism();
+		else if ( executorService instanceof SequentialExecutorService )
+			return ( ( SequentialExecutorService ) executorService ).getParallelism();
+		return Runtime.getRuntime().availableProcessors();
```

Comment:
this `instanceof` cascade could be done once on construction to create a reusable `IntSupplier`

Reply:
Done. I also measure performance, and it reduces execution time by two.

Reply:
Sorry it the opposite. I red the performance measurement results wrongly. Using `IntSupplier` results in less performance. So I removed it again.